### PR TITLE
Set permissions for files and directories created in PGDATA

### LINKED ISSUE #1899: allow-group-access clause in initdb seems not woking correctly
**Describe the bug**
When using allow-group-access clause in initdb still have some files stlil created with 600 permisstions
```
find /data/postgresql/11/data/ -type f  -perm 600
/data/postgresql/11/data/patroni.dynamic.json
/data/postgresql/11/data/postgresql.base.conf.backup
/data/postgresql/11/data/postgresql.conf.backup
/data/postgresql/11/data/recovery.done
```
As far as I know first 3 files are definitely created by patroni. recovery.* file also appears after restarting patroni daemon.

**To Reproduce**
1. Create patroni config with allow-group-access option
```
initdb:
    - allow-group-access
```
2. (Re)initialize cluster.
3. Observe file permissions in data directory.

**Expected behavior**
All files mast have group access permissions

**Environment**
- Patroni version: patroni 1.6.5
- PostgreSQL version: PostgreSQL 11.8 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0, 64-bit
- DCS (and its version): etcd Version: 3.2.17

I am not allowed to upgrade PostgreSQL and Patroni versions right now.

**Patroni configuration file**
Not sure it's needed. Please let me know if it is.

**patronictl show-config**
```
patronictl -c /etc/patroni/config.yml show-config
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    max_connections: '200'
    max_locks_per_transaction: '64'
    max_prepared_transactions: '3'
    max_worker_processes: '1'
    track_commit_timestamp: false
    wal_level: replica
    wal_log_hints: true
  use_pg_rewind: true
  use_slots: true
retry_timeout: 10
ttl: 30
```

**Have you checked Patroni logs?**
Not sure it's needed. Please let me know if it is.

**Have you checked PostgreSQL logs?**
Not sure it's needed. Please let me know if it is.

**Have you tried to use GitHub issue search?**
Was unable to find anything usefull for allow-group-access

### PR DESCRIPTION
**Pull request by CyberDem0n**
Postgres supports two types of permissions:
1. owner only
2. group readable

By default the first one is used because it provides better security. But, sometimes people want to run a backup tool with the user that is different from postgres. In this case the second option becomes very useful. Unfortunately it didn't work correctly because Patroni was creating files with owner access only permissions.

This PR changes the behaviour and permissions on files and directories that are created by Patroni will be calculated based on permissions of PGDATA. I.e., they will get group readable access when it is necessary.

Close #1899
Close #1901

### CODE REVIEW

File: patroni/postgresql/config.py [hunk 1]
```diff
@@ -438,7 +448,9 @@ def write_postgresql_conf(self, configuration: Optional[CaseInsensitiveDict] = N
     def append_pg_hba(self, config: List[str]) -> bool:
         if not self.hba_file and not self._config.get('pg_hba'):
             with open(self._pg_hba_conf, 'a') as f:
+                self.set_file_permissions(self._pg_hba_conf)
                 f.write('\n{}\n'.format('\n'.join(config)))
+            self.set_file_permissions(self._pg_hba_conf)
```
Comment (barthisrael) [inline patroni/postgresql/config.py:453]:
```suggestion
                f.write('\n{}\n'.format('\n'.join(config)))
            self.set_file_permissions(self._pg_hba_conf)
```

I guess we can remove one of the apparently redundant calls.

File: patroni/postgresql/config.py [hunk 2]
```diff
@@ -410,6 +419,7 @@ def write_postgresql_conf(self, configuration: Optional[CaseInsensitiveDict] = N
             configuration['hot_standby_feedback'] = 'on'
 
         with ConfigWriter(self._postgresql_conf) as f:
+            self.set_file_permissions(self._postgresql_conf)
```
Comment (barthisrael) [inline patroni/postgresql/config.py:422]: Just a side thought...

I was wondering if somehow we could use `os.fchmod` inside `ConfigWriter.__enter__` to set permissions of the file descriptor as we are always calling `self.set_file_permissions` right after entering the context.

However that class is not "PGDATA aware", and we would somehow copy the logic of `set_file_permissions`, so...

File: tests/test_file_perm.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        mock_stat.return_value.st_mode = stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP
+        pg_perm.set_permissions_from_data_directory('test')
+        self.assertEqual(mock_umask.call_args[0][0], stat.S_IWGRP | stat.S_IRWXO)
+        self.assertEqual(mock_logger.call_args[0][0], 'Can not set umask to %03o: %r')
+
+    @patch('os.stat', Mock(side_effect=FileNotFoundError))
+    @patch('patroni.file_perm.logger.error')
+    def test_set_permissions_from_data_directory(self, mock_logger):
+        pg_perm.set_permissions_from_data_directory('test')
+        self.assertEqual(mock_logger.call_args[0][0], 'Can not check permissions on %s: %r')
```
Comment (barthisrael) [inline tests/test_file_perm.py:25]: What about adding a couple tests to check if `__set_group_permissions` or `__set_owner_permissions` is being called?

**Review by barthisrael — CHANGES_REQUESTED** [review]
The PR looks good overall. Posted a few comments about docstrings, and a few side thoughts.

File: patroni/file_perm.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            self.__set_umask()
+
+    @property
+    def dir_create_mode(self) -> int:
+        """Directory permissions."""
+        return self.__pg_dir_create_mode
+
+    @property
+    def file_create_mode(self) -> int:
+        """File permissions"""
```
Comment (hughcapet) [inline patroni/file_perm.py:81]:
```suggestion
        """File permissions."""
```

File: tests/test_postgresql.py [hunk 5]
```diff
@@ -523,8 +523,9 @@ def test_sysid(self):
     def test_save_configuration_files(self):
         self.p.config.save_configuration_files()
 
-    @patch('os.path.isfile', Mock(side_effect=[False, True]))
-    @patch('shutil.copy', Mock(side_effect=IOError))
+    @patch('os.path.isfile', Mock(side_effect=[False, True, False, True]))
+    @patch('shutil.copy', Mock(side_effect=[None, IOError]))
+    @patch('os.chmod', Mock())
```
Comment (hughcapet) [inline tests/test_postgresql.py:528]: I guess, perfectly, the tests should also check that `chmod`  was actually called with the proper permissions everywhere it is required (where it was added in this PR) 😇

File: tests/test_file_perm.py (hunk 3 above)
Comment (CyberDem0n) [reply tests/test_file_perm.py:25]: Well, `umask` could only called when `os.stat()` didn't raise any exception.
Therefore one or another is successfully called :)

Comment (barthisrael) [reply tests/test_file_perm.py:25]: Ah, sure. I guess I didn't express myself clearly, sorry.
I meant based on a few sample mock values for `st_mode`, to check if it's taking the expected code path.

Comment (CyberDem0n) [reply tests/test_file_perm.py:25]: At least one path is checked here: https://github.com/zalando/patroni/pull/2781/files#diff-6251748fa4c45a038329a6f9eb925a57f7a1c07b50b64823ca36ea20804748efR16-R18

Comment (barthisrael) [reply tests/test_file_perm.py:25]: Right, the group one.
Your call on adding a test for the other or not.

**Comment by hughcapet** [comment]
👍

**Pull request merged**